NeoML/src/Dnn/Layers/DnnHeadAdapterLayer.cpp

```diff
@@ -0,0 +1,238 @@
+
+	setInternalDnnParams();
+	calcBlobsForBackwardAndLearn();
+	outputDescs[0] = internalDnn->GetLayer("sink")->inputDescs[0];
+}
+
+void CDnnHeadAdapterLayer::RunOnce()
+{
+	innerInputBlobs.DeleteAll();
+	NeoAssert(inputBlobs.Size() == 1);
+
+	auto headSource = static_cast<CCompositeSourceLayer*>(internalDnn->GetLayer("source").Ptr());
+	headSource->SetBlob(inputBlobs[0]);
+	internalDnn->runOnce( 0 );
+
+	auto headSink = static_cast<CCompositeSinkLayer*>(internalDnn->GetLayer("sink").Ptr());
+	outputBlobs[0] = headSink->GetInputBlob()->GetCopy();
+
+	// save current inputs for next Backward
+	if(IsBackwardNeeded() || IsLearningEnabled()) {
+		rememberInputs();
+	}
+}
+
+void CDnnHeadAdapterLayer::BackwardOnce()
+{
+	NeoAssert(internalDnn != 0);
+	NeoAssert(internalDnn->isBackwardPerformed == GetDnn()->isBackwardPerformed);
+
+	if(IsBackwardNeeded()) {
+		auto headSource = static_cast<CCompositeSourceLayer*>(internalDnn->GetLayer("source").Ptr());
+		headSource->SetDiffBlob(inputDiffBlobs[0]);
+	}
+
+	// getting diffs from external dnn
+	auto headSink = static_cast<CCompositeSinkLayer*>(internalDnn->GetLayer("sink").Ptr());
+	headSink->SetDiffBlob(outputDiffBlobs[0]);
+
+	// loading inputs from last RunOnce (may be overwritten by other AdapterLayer)
+	loadInputs();
+
+	internalDnn->backwardRunAndLearnOnce( 0 );
+}
+
+void CDnnHeadAdapterLayer::LearnOnce()
+{
+	if(!IsBackwardPerformed()) {
+		BackwardOnce();
+	}
+}
+
+void CDnnHeadAdapterLayer::SetDnnHead(CPtr<CHeadStruct> _head)
+{
+	head = _head;
+	internalDnn = head->dnn;
+	num = head->connections.Size();
+	head->connections.Add(this);
+	ForceReshape();
+}
+
+CHeadStruct* CDnnHeadAdapterLayer::GetDnnHead()
+{
+	return head;
+}
+
+void CDnnHeadAdapterLayer::setInternalDnnParams()
+{
+	NeoAssert(internalDnn != 0);
+	// If the backward pass requirements have changed, call reshape
+	bool forcedReshape = internalDnn->IsBackwardPerformed() != GetDnn()->IsBackwardPerformed();
+
+	// Set the internal network parameters from the external network parameters
+	internalDnn->setProcessingParams(GetDnn()->IsRecurrentMode(), GetDnn()->GetMaxSequenceLength(),
+		GetDnn()->IsReverseSequense(), GetDnn()->IsBackwardPerformed());
+	internalDnn->RequestReshape(forcedReshape);
+	internalDnn->SetInitializer(GetDnn()->GetInitializer());
+
+	if(IsLearningEnabled()) {
+		internalDnn->EnableLearning();
+	} else {
+		internalDnn->DisableLearning();
+	}
+
+	bool needLearn = false;
+	for(int i = 0; i < internalDnn->layers.Size(); ++i) {
+		needLearn |= internalDnn->layers[i]->IsLearningPerformed();
+		auto layer = dynamic_cast<CDropoutLayer*>(internalDnn->layers[i].Ptr());
+		if(layer != nullptr) {
+			layer->SetHeadCounter(head->connections.Size());
+		}
+	}
+
+	// Switch learning on or off
+	if(needLearn) {
+		EnableLearning();
+	} else {
+		DisableLearning();
+	}
+
+	internalDnn->SetSolver(GetDnn()->GetSolver());
+	internalDnn->reshape();
+}
+
+void CDnnHeadAdapterLayer::rememberInputs()
```

Comment:
Maybe better naming, something like
`save / reload`
`retain / recover`
`reserve / reinit`

'store/load' used in searialization to file